### DISCUSSION TITLE
Performance of Annotations

### DISCUSSION THREAD

**Discussion by justinsmalley** [grafana/grafana#58781] [Dashboards, Panels & Visualizations]
For a particular dashboard I am using there are about 15 panels, and 8 different types of queried annotations. These are queries to influxdb or postgres to add extra metadata to our monitored assets. Depending on the annotation type, there can be as many as 1 per minute, and as few as zero. When we have multiple the annotations enabled there is a noticeable slowdown in the dashboard performance. When using the dashboard, sometimes I am looking at a time span of only a few minutes, and sometimes up to 6 months. The sweet spot I found was to limit each query for annotations to a max of 200, otherwise the dashboard responsiveness would grind to a halt. Note that since these are queried annotations, they appear on most panels in the dashboard. I am using mostly time-series and state timeline panels, on which both show up.

From some testing, the annotations appear to be extremely inefficient with memory. I don't know if this is duplicating data per panel in the background, or actually rendering the shapes. Here are some benchmarks I took for the browser page. They were all for the same dashboard, variable settings, and 90 day time range.

- 53 MB: All annotations off, all panels deleted except 1
- 60 MB: All annotations off, all panels minimized except 1
- 78 MB: All annotations off, thirteen most common panels open
- 165 MB: 1 annotation enabled (limit 200)
- 119 MB: Annotations turned back off
- 437 MB: 4 types of annotations enabled (limit 200 each)
- **1221 MB**: 8 types of annotations enabled (limit 200 each)
- 112 MB: all annotations turned back off

On a computer with 64GB memory, it was still mostly responsive, but on another computer with 16GB memory it was very slow to point of being unusable. If anyone in the community has any tips to improve the responsiveness through configuration I'd be happy to hear.

What I (and hopefully other users) would like to see are some general enhancements to annotations to help improve the responsiveness in some of these heavy load cases, as well as some more general usability 

1. Investigate and determine if there is a way to improve general responsiveness and performance by optimizing annotation memory usage, rendering, or both.
2. Currently when an annotation is toggled on or off, ALL panels queries are re-executed. There is no need to do this from a data perspective, since the underlying data cannot have changed.
3. Show annotation colour in the variable toggle button at the top. This could be done by adding a coloured pill shape (like in the time series panel legend), or some other method to show the colour.
4. Thickness of the annotation line: In the old Graph panel, it was significantly thinner (1 pixel?), whereas in the new time series panel it is much wider. In my opinion, this the thinner annotation line was more visually pleasing an obscured less of the data behind it. To address this, I think an option for line thickness and possibly opacity would be helpful.
5. When multiple annotations that land on the same pixel, the "bottom" one is not visible in the popup. [already reported](https://github.com/grafana/grafana/issues/53359) as an issue.

Thanks!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
